docs/source/indices.rst

```diff
@@ -0,0 +1,176 @@
+.. Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht
+
+   Distributed under the terms of the BSD 3-Clause License.
+
+   The full license is in the file LICENSE, distributed with this software.
+
+Indices
+=======
+
+Definition
+----------
+
+There are two types of indices: *array indices* and *flat indices*. Consider this example (stored in row-major):
+
+.. code-block:: cpp
+
+    #include <xtensor/xtensor.hpp>
+    #include <xtensor/xarray.hpp>
+    #include <xtensor/xio.hpp>
+
+    int main()
+    {
+        xt::xarray<size_t> a = xt::arange<size_t>(3*4);
+
+        a.reshape({3,4});
+
+        std::cout << a << std::endl;
+    }
+
+Which prints
+
+.. code-block:: none
+
+    {{ 0,  1,  2,  3},
+     { 4,  5,  6,  7},
+     { 8,  9, 10, 11}}
+
+The *array index* ``(1,2)`` corresponds to the *flat index* ``6``.
+
+Array indices
+-------------
+
+Functions like ``xt::argwhere(a < 5)`` return a ``std::vector`` of *array indices*. Using the same matrix as above, we can do
+
+.. code-block:: cpp
+
+    int main()
+    {
+        xt::xarray<size_t> a = xt::arange<size_t>(3*4);
+
+        a.reshape({3,4});
+
+        auto idx = xt::from_indices(xt::argwhere(a >= 6));
+
+        std::cout << idx << std::endl;
+    }
+
+which prints
+
+.. code-block:: none
+
+    {{1, 2},
+     {1, 3},
+     {2, 0},
+     {2, 1},
+     {2, 2},
+     {2, 3}}
+
+Here we observe that to work print we need to convert the ``std::vector`` to a ``xt::xtensor<size_t, 2>`` array, which is done using ``xt::from_indices``.
+
+From array indices to flat indices
+----------------------------------
+
+To convert the array indices to a ``xt::xtensor<size_t, 1>`` of flat indices, ``xt::ravel_indices`` can be used. For to same example:
+
+.. code-block:: cpp
+
+    #include <xtensor/xtensor.hpp>
+    #include <xtensor/xarray.hpp>
+    #include <xtensor/xio.hpp>
+
+    int main()
+    {
+        xt::xarray<size_t> a = xt::arange<size_t>(3*4);
+
+        a.reshape({3,4});
+
+        auto idx = xt::ravel_indices(xt::argwhere(a >= 6), a.shape());
+
+        std::cout << idx << std::endl;
+    }
+
+which prints
+
+.. code-block:: none
+
+    { 6,  7,  8,  9, 10, 11}
+
+.. note::
+
+    To convert to a ``std::vector`` use
+
+    .. code-block:: cpp
+
+        auto idx = xt::ravel_indices<xt::ravel_vector_tag>(xt::argwhere(a >= 6), a.shape());
+
+1-d arrays: array indices == flat indices
```

Comment:
Somme nitpicking comments :)

We have generally written `1-D` across the docs, instead of `1-d`.

Also, for formatting in examples, we write generally `3 * 4` instead of `4*3` (spaces around arithmetic operators) and keep spaces afters commas.